### ISSUE TITLE
tasks in guestrecipe report aborted state

### ISSUE THREAD

**Issue by jtluka** [restraint-harness/restraint#294]
In our beaker/restraint setup we hit an issue when tasks running in guestrecipe report aborted state. The problem appears when host executes /virt/start task and later on the host reboots the OS. That results in guest's task run interruption and when guest's restraintd resumes task execution (host starts the guests with /virt/stat again), the abort state is set [1] and reported back to beaker server.

[1] https://github.com/restraint-harness/restraint/blob/c468b3d8284ea8395ac4f1ff11c592a7357f5d5d/src/task.c#L1266

I'm wondering if this can be fixed in some way or whether I need to find a different way how to workaround the problem.

A sample beaker xml that reproduces problem is:
```xml
<job retention_tag="120days" group="networking-CI-LNST">
  <whiteboard>test aborts</whiteboard>
  <recipeSet priority="Urgent">
    <recipe whiteboard="" role="None" ks_meta="method=http harness='restraint-rhts beakerlib-redhat' redhat_ca_cert" kernel_options="ksdevice=bootif" kernel_options_post="isolcpus=0,1,2,5">
      <guestrecipe whiteboard="guest1 - LNST agent" role="None" ks_meta="method=http harness='restraint-rhts beakerlib-redhat' redhat_ca_cert" kernel_options="" kernel_options_post="" guestname="guest1" guestargs="--ram=4096 --vcpus=1 --file-size=20 --kvm --hvm --cpuset=1">
        <autopick random="false"/>
        <watchdog panic="None"/>
        <packages/>
        <ks_appends/>
        <repos/>
        <distroRequires>
                    <and>
                        <distro_family op="=" value="RedHatEnterpriseLinux8"/>
                        <distro_tag op="=" value="RELEASED"/>
                        <variant op="=" value="BaseOS"/>
                        <arch op="=" value="x86_64"/>
                    </and>
                </distroRequires>
        <hostRequires>
          <system_type value="Virtual"/>
        </hostRequires>
        <partitions/>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
      </guestrecipe>
      <guestrecipe whiteboard="guest2" role="None" ks_meta="method=http harness='restraint-rhts beakerlib-redhat' redhat_ca_cert" kernel_options="" kernel_options_post="" guestname="guest2" guestargs="--ram=4096 --vcpus=1 --file-size=20 --kvm --hvm --cpuset=2">
        <autopick random="false"/>
        <watchdog panic="None"/>
        <packages/>
        <ks_appends/>
        <repos/>
        <distroRequires>
                    <and>
                        <distro_family op="=" value="RedHatEnterpriseLinux8"/>
                        <distro_tag op="=" value="RELEASED"/>
                        <variant op="=" value="BaseOS"/>
                        <arch op="=" value="x86_64"/>
                    </and>
                </distroRequires>
        <hostRequires>
          <system_type value="Virtual"/>
        </hostRequires>
        <partitions/>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
      </guestrecipe>
      <autopick random="false"/>
      <watchdog panic="None"/>
      <packages/>
      <ks_appends/>
      <repos/>
      <distroRequires>
                <and>
                    <distro_family op="=" value="RedHatEnterpriseLinux8"/>
                    <distro_tag op="=" value="RELEASED"/>
                    <variant op="=" value="BaseOS"/>
                    <arch op="=" value="x86_64"/>
                </and>
            </distroRequires>
      <partitions/>
      <task name="/virt/install" role="None">
        <params>
          <param name="KILLTIMEOVERRIDE" value="5100"/>
        </params>
      </task>
      <task name="/virt/start" role="None"/>
      <task name="/distribution/command" role="None">
        <params>
          <param name="CMDS_TO_RUN" value="sleep 120"/>
        </params>
      </task>
      <task name="/distribution/kernelinstall" role="None">
        <params>
          <param name="KERNELARGNAME" value="kernel"/>
          <param name="KERNELARGVERSION" value="4.18.0-305.el8"/>
          <param name="KERNELARGVARIANT" value="up"/>
          <param name="KERNELARGEXTRAMODULES" value="1"/>
        </params>
      </task>
      <task name="/distribution/dummy" role="None">
        <params>
          <param name="AVC_ERROR" value="+no_avc_check"/>
        </params>
      </task>
      <task name="/distribution/dummy" role="None">
        <params>
          <param name="AVC_ERROR" value="+no_avc_check"/>
        </params>
      </task>
    </recipe>
    <recipe whiteboard="" role="None" ks_meta="method=http harness='restraint-rhts beakerlib-redhat' redhat_ca_cert" kernel_options="ksdevice=bootif" kernel_options_post="isolcpus=0,1,2,5">
      <guestrecipe whiteboard="guest1 - LNST agent" role="None" ks_meta="method=http harness='restraint-rhts beakerlib-redhat' redhat_ca_cert" kernel_options="" kernel_options_post="" guestname="guest1" guestargs="--ram=4096 --vcpus=1 --file-size=20 --kvm --hvm --cpuset=1">
        <autopick random="false"/>
        <watchdog panic="None"/>
        <packages/>
        <ks_appends/>
        <repos/>
        <distroRequires>
                    <and>
                        <distro_family op="=" value="RedHatEnterpriseLinux8"/>
                        <distro_tag op="=" value="RELEASED"/>
                        <variant op="=" value="BaseOS"/>
                        <arch op="=" value="x86_64"/>
                    </and>
                </distroRequires>
        <hostRequires>
          <system_type value="Virtual"/>
        </hostRequires>
        <partitions/>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
      </guestrecipe>
      <guestrecipe whiteboard="guest2" role="None" ks_meta="method=http harness='restraint-rhts beakerlib-redhat' redhat_ca_cert" kernel_options="" kernel_options_post="" guestname="guest2" guestargs="--ram=4096 --vcpus=1 --file-size=20 --kvm --hvm --cpuset=2">
        <autopick random="false"/>
        <watchdog panic="None"/>
        <packages/>
        <ks_appends/>
        <repos/>
        <distroRequires>
                    <and>
                        <distro_family op="=" value="RedHatEnterpriseLinux8"/>
                        <distro_tag op="=" value="RELEASED"/>
                        <variant op="=" value="BaseOS"/>
                        <arch op="=" value="x86_64"/>
                    </and>
                </distroRequires>
        <hostRequires>
          <system_type value="Virtual"/>
        </hostRequires>
        <partitions/>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
        <task name="/distribution/command" role="None">
          <params>
            <param name="CMDS_TO_RUN" value="sleep 120"/>
          </params>
        </task>
      </guestrecipe>
      <watchdog panic="None"/>
      <packages/>
      <ks_appends/>
      <repos/>
      <distroRequires>
                <and>
                    <distro_family op="=" value="RedHatEnterpriseLinux8"/>
                    <distro_tag op="=" value="RELEASED"/>
                    <variant op="=" value="BaseOS"/>
                    <arch op="=" value="x86_64"/>
                </and>
            </distroRequires>
      <partitions/>
      <task name="/virt/install" role="None">
        <params>
          <param name="KILLTIMEOVERRIDE" value="5100"/>
        </params>
      </task>
      <task name="/virt/start" role="None"/>
      <task name="/distribution/command" role="None">
        <params>
          <param name="CMDS_TO_RUN" value="sleep 120"/>
        </params>
      </task>
      <task name="/distribution/kernelinstall" role="None">
        <params>
          <param name="KERNELARGNAME" value="kernel"/>
          <param name="KERNELARGVERSION" value="4.18.0-305.el8"/>
          <param name="KERNELARGVARIANT" value="up"/>
          <param name="KERNELARGEXTRAMODULES" value="1"/>
        </params>
      </task>
      <task name="/distribution/dummy" role="None">
        <params>
          <param name="AVC_ERROR" value="+no_avc_check"/>
        </params>
      </task>
      <task name="/distribution/dummy" role="None">
        <params>
          <param name="AVC_ERROR" value="+no_avc_check"/>
        </params>
      </task>
    </recipe>
  </recipeSet>
</job>
```

**Comment by wackrat** [comment]
If you want the host running kernel `4.18.0-305.el8` while the guests are active, the recipe should run the `/distribution/kernelinstall` task before it runs the `/virt/start` task.

If you want the host running a different kernel while the guests are active, then the `/distribution/kernelinstall` task serves no apparent purpose in those sample recipes, although perhaps the job XML here is not fully representative of the intent of the host recipe, if it's a simplified sample. 

If you want the guest recipes to finish before the  `/distribution/kernelinstall` task reboots the host, that would presumably require some sort of coordination between the host and guest recipes.

If you specifically want the guest recipes to span a host reboot, that might require writing a new task or modifying an existing one, or suspending the guests before the host reboot and resuming them after the host reboot.

I noticed a couple of things in a [recipe](https://beaker.engineering.redhat.com/recipes/13805757#task159389823) I found which, I think, matches the job XML here.

One thing I noticed is that each successive invocation of `/distribution/command` repeats the installation of the `distribution-distribution-command` package, as seen in all but one of the `harness.log` files.

The other thing I noticed is that the one task in that recipe which aborted reported the following error in its `taskout.log`.
```
make: *** No rule to make target 'run'.  Stop.
```
That seems consistent with the premise that the task `Makefile` does not survive a reboot, although it's possible that the task restart also somehow lost track of which directory it's supposed to run in. The next invocation of `/distribution/command` does not have that problem, presumably because the `distribution-distribution-command` installation reported in its `harness.log` creates the `Makefile`.

I also noticed that there is no apparent support for handling a reboot in the `/distribution/command` task.

If the intention behind this sample job is to allow guest recipes to span host reboots, I recommend suspend/resume of each guest on the host before and after each host reboot. It seems the only alternative might be confining the guest recipes to tasks which properly handle reboots (including crashes, which is what host reboots without suspend/resume amount to).

**Comment by jtluka** [comment]
@wackrat I've tested your suggestion and it works fine. I'll close this issue.

**Comment by wackrat** [comment]
I also tested my suggestion (before I noticed the replies here), and I found a minor limitation.

If the different kernel versions on the host aren't sufficiently compatible, `virsh restore` can fail after a host reboot, with a file saved by `virsh save`. The specific incompatibility I encountered was due to a kernel bug (on the host) which garbled the guest registers. Part of the error message from `virsh restore` follows.
```
qemu-kvm: error while loading state for instance 0x0 of device 'cpu'
```
I wouldn't expect this to be an issue in the absence of kernel bugs on the host related to KVM, but there might be other compatibility issues I haven't pondered. Still, I'd expect this approach to work well with two sufficiently similar kernel versions on the host.

**Comment by wackrat** [comment]
I discovered another variant of the limitation mentioned here.

In particular, I saw the following error from `virsh restore` after booting a modified `4.18.0-361.el8` on the host while attempting to restore a KVM guest saved after booting a modified `4.18.0-425.13.1.el8_7` on the host.

```
qemu-kvm: Invalid value 238 expecting positive value <= 236
qemu-kvm: Failed to load cpu:cpreg_vmstate_array_len
qemu-kvm: error while loading state for instance 0x0 of device 'cpu'
```

Each modified kernel had a fix for the kernel bug mentioned here. The last of those three lines matches the error message mentioned here previously; the other two provide additional detail not present in prior failures.

Reversing the boot order of the two kernels does not allow `virsh restore` with the newer kernel to succeed on a file saved with the older kernel, so it does not seem to be a question of kernel upgrade versus kernel downgrade, for the host reboot; the two host kernels need to have a compatible notion of the guest registers, buggy or not. The additional details (the ones that mention `cpu:cpreg_vmstate_array_len`) appear only with the kernel downgrade, not the kernel upgrade.
